### PR TITLE
OPA Taxonomy

### LINKED ISSUE #536: OPA Governance Input Taxonomy
Depend on

Task list:

- [ ] Wait for merge of #475 
- [ ] Finalize the OPA input fields in structural taxonomy part
- [ ] Finalize the corresponding  OPA values in values taxonomy part
- [ ] Test the OPA taxonomy by writing corresponding test cases
- [ ] Validate the taxonomy in OPA (both structure-based and value-based) at different control points in OPA connector

### PR DESCRIPTION
**Pull request by ritwikc**
Signed-off-by: Ritwik <ritwik@datameshproj.sl.cloud9.ibm.com>

Partially fixes #536.

### CODE REVIEW

File: config/taxonomy/opa.structs.schema.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
+{
+    "$schema": "http://json-schema.org/draft-04/schema#",
+    "$id": "opa.schema.json",
+    "title": "OPA input Taxonomy",
```
Comment (simanadler) [inline config/taxonomy/opa.structs.schema.json:4]: OPA Data Request Taxonomy

File: config/taxonomy/opa.structs.schema.json [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    "description": "This is the OPA taxonomy.",
```
Comment (simanadler) [inline config/taxonomy/opa.structs.schema.json:5]: Something a little more meaningful here would be good.  Ex: This taxonomy defines the structure sent to Open Policy Agent policy engine, which determines whether or how the data requested may be used by the requester.

File: config/taxonomy/opa.structs.schema.json [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    "definitions": {
```
Comment (simanadler) [inline config/taxonomy/opa.structs.schema.json:6]: I don't think these definition are needed.

File: config/taxonomy/opa.structs.schema.json [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+        "opaGovernanceRequest":{"$ref":"policymanager.structs.schema.json#/definitions/governance_decision_request"},
+        "opaGovernanceResponse":{"$ref":"policymanager.structs.schema.json#/definitions/governance_decision_response"}
+        },
+    "properties": {
+        "opaGovernanceRequest": {"$ref": "#/definitions/opaGovernanceRequest"},
```
Comment (simanadler) [inline config/taxonomy/opa.structs.schema.json:11]: Why not directly reference the external file?

File: config/taxonomy/opa.structs.schema.json [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+        "opaGovernanceResponse": {"$ref": "#/definitions/opaGovernanceResponse"}
```
Comment (simanadler) [inline config/taxonomy/opa.structs.schema.json:12]: Why not directly reference the external file?

**Review by simanadler — CHANGES_REQUESTED** [review]

File: config/taxonomy/module.structs.schema.json [hunk 6]
```diff
@@ -4,38 +4,38 @@
...
                     "items": {
-                        "type":"string"
+                        "type": "string"
                     }
-                },
-                "level": {
-                    "type": "object",
+                }
+            },
+            "oneOf": [{
```
Comment (roee88) [inline config/taxonomy/module.structs.schema.json:23]: oneOf is validation. It should be in values not in structs.

File: config/taxonomy/module.values.schema.json [hunk 7]
```diff
@@ -9,33 +9,27 @@
...
-            "description": "Transformation that can be performed on data.",
+        "protocol": {
             "type": "string",
-            "enum": ["mask", "redact", "encrypt"]
-        },
-        "data_level_name": {
-            "type":"string",
-            "enum":["UNKNOWN", "DATASET", "COLUMN", "ROW", "CELL"]
+            "description": "The mechanism via which the data will be accessed.",
+            "enum": ["s3", "kafka", "db2-jdbc", "m4d-arrow-flight"]
```
Comment (roee88) [inline config/taxonomy/module.values.schema.json:15]: How is this related to this PR? Will conflict with the work @eletonia is doing.

File: config/taxonomy/module.values.schema.json [hunk 8]
```diff
@@ -9,33 +9,27 @@
...
         },
-        "data_level_code": {
-            "type":"integer",
-            "description":"0=unknown, 1=dataset, 2=column, 3=row, 4=cell",
-            "minimum": 0,
-            "maximum": 4          
+        "action_name_columns": {
```
Comment (roee88) [inline config/taxonomy/module.values.schema.json:17]: Future: pick a better name than action_name_columns

File: config/taxonomy/module.values.schema.json [hunk 9]
```diff
@@ -9,33 +9,27 @@
...
+            "type": "string",
+            "description": "Type of actions possible on columns",
+            "enum": ["RemoveColumn", "RedactColumn", "EncryptColumn", "PeriodicBlackout"]
         },
-        "protocol": {
-            "type":"string",
-            "description": "The mechanism via which the data will be accessed.",
-            "enum":["s3", "kafka", "db2-jdbc", "m4d-arrow-flight"]
+        "action_name_dataset": {
```
Comment (roee88) [inline config/taxonomy/module.values.schema.json:22]: Future: pick a better name than action_name_dataset

File: config/taxonomy/policymanager_response.structs.schema.json [hunk 10]
```diff
@@ -0,0 +1,29 @@
+{
+    "$schema": "http://json-schema.org/draft-04/schema#",
+    "$id": "opa.schema.json",
```
Comment (roee88) [inline config/taxonomy/policymanager_response.structs.schema.json:3]: wrong name

File: config/taxonomy/policymanager_response.structs.schema.json [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+    "title": "OPA Data Response Taxonomy",
+    "description": "This taxonomy defines the structure of response received from Policy Agent policy engine with enforcements.",
```
Comment (roee88) [inline config/taxonomy/policymanager_response.structs.schema.json:5]: change description because this isn't OPA specific

File: config/taxonomy/policymanager_response.structs.schema.json [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+        "result": {
+            "type": "array",
+            "description": "While showing the result, action contains the action type and the associated entity on which action has been taken.",
+            "items": {
+                "properties": {
+                    "action": { "$ref": "module.structs.schema.json#/definitions/action" },
+                    "used_policy": {
+                        "description": "The list of policies on which the decision was based.",
+                        "type": "string"
+                    }
```
Comment (roee88) [inline config/taxonomy/policymanager_response.structs.schema.json:21]: This is different than the design doc because it would require rules to look like:
```rego
rule[{"used_policy": "Redact PII columns for CBA", "action": {"name": "RedactColumn", "columns": columns}] {
  ...
}
```

Instead of:
```rego
rule[{"description": "Redact PII columns for CBA", "action": "RedactColumn", "columns": columns}] {
  ...
}
```

The differences are:
1. `used_policy` instead of `description` but we already discussed it and said that the change is fine. I would consider changing it to "policy" instead of "used_policy" though.
2. `action` is a structure instead of being flat

I'm not asking for a change but rather for you to consider if this is what is desired in terms of how people would write opa policies. Thanks.

**Review by roee88 — COMMENTED** [review]
I tried to focus on changes related to policy manager response taxonomy

File: config/taxonomy/catalog.structs.schema.json [hunk 13]
```diff
@@ -28,10 +28,14 @@
             "description": "A grouping of discrete geographies that are not necessarily in a common hierarchy.",
             "items": {
                 "properties": {
-                    "geography": {"$ref": "#/definitions/geography"}
+                    "geography": { "$ref": "#/definitions/geography" }
                 }
             }
         },
+        "tag": {
```
[the remark this answers is not in the document]
Comment (ritwikc) [reply config/taxonomy/catalog.structs.schema.json]: Done

File: config/taxonomy/catalog.structs.schema.json [hunk 14]
```diff
@@ -51,20 +55,48 @@
...
+                        },
+                        "updatedBy": {
+                            "type": "string",
+                            "description": "Indicates who updted the asset"
                         },
-                        "required":["name"]
+                        "totalRatings": {
+                            "type": "string",
+                            "description": "Indicates the total rating of the asset"
+                        }
```
[the remark this answers is not in the document]
Comment (ritwikc) [reply config/taxonomy/catalog.structs.schema.json]: Included in additionalProperties.

File: config/taxonomy/module.structs.schema.json [hunk 15]
```diff
@@ -4,38 +4,38 @@
...
+        "action": {
             "type": "object",
-            "description": "Transformation that may be applied to data.",
+            "description": "Contains the details about action and on which entity (columns, dataset) the action is enforced",
             "properties": {
-                "name" : {"$ref": "module.values.schema.json#/definitions/transformation_name"},
-                "params": {
+                "name": {
+                    "type": "string",
+                    "description": "Represents the action type"
```
[the remark this answers is not in the document]
Comment (ritwikc) [reply config/taxonomy/module.structs.schema.json]: Added in the description

File: config/taxonomy/module.values.schema.json (hunk 7 above)
Comment (ritwikc) [reply config/taxonomy/module.values.schema.json]: Removed protocol to avoid conflict with Tonia.

File: config/taxonomy/module.values.schema.json (hunk 8 above)
Comment (ritwikc) [reply config/taxonomy/module.values.schema.json]: Changed the name to allowable_action_columns

File: config/taxonomy/module.values.schema.json (hunk 9 above)
Comment (ritwikc) [reply config/taxonomy/module.values.schema.json]: Changed the name to allowable_action_dataset

File: config/taxonomy/policymanager_response.structs.schema.json (hunk 10 above)
Comment (ritwikc) [reply config/taxonomy/policymanager_response.structs.schema.json]: Corrected it.

File: config/taxonomy/policymanager_response.structs.schema.json (hunk 11 above)
Comment (ritwikc) [reply config/taxonomy/policymanager_response.structs.schema.json]: Changed the description

File: config/taxonomy/policymanager_response.structs.schema.json (hunk 12 above)
Comment (ritwikc) [reply config/taxonomy/policymanager_response.structs.schema.json]: Changed used_policy to policy. For this I prefer to keep it in the format:
rule[{"used_policy": "Redact PII columns for CBA", "action": {"name": "RedactColumn", "columns": columns}] {
  ...
}
Because, an action here us corresponds to the name of that action and the columns on which the action has been made. Seems to be a clearer view of the action. @roee88

**Review by roee88 — APPROVED** [review]
There are some things here that wouldn't pass the taxonomy validation tool, but it's not fair to request changes before the tool is ready so we'll revisit it later. There are also some open questions on the structure. I suggest to merge this as is and iterate because it's blocked for too long.

**Comment by simanadler** [comment]
> Signed-off-by: Ritwik [ritwik@datameshproj.sl.cloud9.ibm.com](mailto:ritwik@datameshproj.sl.cloud9.ibm.com)
> 
> Partially fixes #536.

It would be helpful to reviewers if you indicated which parts of the issue it fixes.

File: config/taxonomy/catalog.structs.schema.json [hunk 16]
```diff
@@ -51,20 +55,42 @@
...
-            "additionalProperties": false
-        }
-    },
+            "required": ["name"],
+            "additionalProperties": {
+                "creationTime": {
+                    "type": "string",
+                    "description": "Time when the asset was created"
+                },
+                "CreatedBy": {
```
Comment (simanadler) [inline config/taxonomy/catalog.structs.schema.json:70]: All the other properties start with a small letter, so I assume this one should as well

File: config/taxonomy/catalog.structs.schema.json [hunk 17]
```diff
@@ -51,20 +55,42 @@
...
+                    "type": "string",
+                    "description": "Owner Id for the user who created the asset"
+                },
+                "updationTime": {
```
Comment (simanadler) [inline config/taxonomy/catalog.structs.schema.json:74]: updation is not a word I don't think.  Did you mean updateTime?

File: config/taxonomy/catalog.structs.schema.json [hunk 18]
```diff
@@ -51,20 +55,42 @@
...
+                    "type": "string",
+                    "description": "Time when the dataset got updated"
+                },
+                "updatedBy": {
+                    "type": "string",
+                    "description": "Indicates who updted the asset"
+                },
+                "totalRatings": {
+                    "type": "string",
+                    "description": "Indicates the total rating of the asset"
```
Comment (simanadler) [inline config/taxonomy/catalog.structs.schema.json:84]: Are these ratings provided by the data users?  If so, please state.  Also, do we assume all catalogs will have this?

File: config/taxonomy/catalog.structs.schema.json [hunk 19]
```diff
@@ -51,20 +55,42 @@
                                 "type": "string",
                                 "description": "name of column or component"
                             },
-                            "tags": { "$ref": "catalog.values.schema.json#/definitions/valid_metadata"}
```
Comment (simanadler) [inline config/taxonomy/catalog.structs.schema.json:54]: By removing the validation of the metadata aren't we are removing one of the key things the taxonomy is meant to address?  I.e. Ensuring that the metadata used in the catalog is the same metadata used in the policies?

**Comment by ritwikc** [comment]
@simanadler In this PR, policymanager_request and policymanager_repsonse schemas have been created. For this, there are some updates/changes made to other existing schema as well. 
Also the corresponding go tests have been implemented.
As a next step in a new PR, I will put these tests in the appropriate places in policy manager so that the validations can be done.

File: config/taxonomy/policymanager.values.schema.json [hunk 20]
```diff
@@ -4,7 +4,7 @@
     "title": "Policy Manager Values Taxonomy",
```
Comment (simanadler) [inline config/taxonomy/policymanager.values.schema.json:4]: These are values for the policy manager **request** values. Name of file and title should be changed accordingly

File: config/taxonomy/catalog.structs.schema.json (hunk 16 above)
Comment (roee88) [reply config/taxonomy/catalog.structs.schema.json:70]: @ritwikc please remove the entire additionalProperties here. We discussed it last time and I was going to leave it for the next iteration but obviously it adds unnecessary overhead in this PR.

File: config/taxonomy/catalog.structs.schema.json (hunk 19 above)
Comment (ritwikc) [reply config/taxonomy/catalog.structs.schema.json:54]: With this update, what we tried to achieve was to make tags as a map so that anything with a key,value pair can appear as a tag. This was actually suggested by @roee88 . Hence, that change was brought in. @simanadler

Comment (roee88) [reply config/taxonomy/catalog.structs.schema.json:54]: Good catch. I suggest to move tags definition to catalog.structs.schema.json.

File: config/taxonomy/catalog.structs.schema.json (hunk 18 above)
Comment (ritwikc) [reply config/taxonomy/catalog.structs.schema.json:84]: These are metadata obtained from a particular asset. So, yes it is given by the data user as and when he is using the asset. This field is present in metadata obtained from both WKC and Egeria as well.

File: config/taxonomy/catalog.structs.schema.json (hunk 17 above)
Comment (ritwikc) [reply config/taxonomy/catalog.structs.schema.json:74]: Yes. Let me correct it.

File: config/taxonomy/catalog.structs.schema.json (hunk 16 above)
Comment (ritwikc) [reply config/taxonomy/catalog.structs.schema.json:70]: Let me correct it.

File: config/taxonomy/policymanager.values.schema.json (hunk 20 above)
Comment (ritwikc) [reply config/taxonomy/policymanager.values.schema.json:4]: Let me correct it.

File: config/taxonomy/catalog.structs.schema.json (hunk 19 above)
Comment (ritwikc) [reply config/taxonomy/catalog.structs.schema.json:54]: @roee88 Do you suggest to keep the tags as map in catalog.structs.schema.json as it is now ?

File: config/taxonomy/catalog.structs.schema.json (hunk 16 above)
Comment (ritwikc) [reply config/taxonomy/catalog.structs.schema.json]: Done

File: config/taxonomy/catalog.structs.schema.json (hunk 17 above)
Comment (ritwikc) [reply config/taxonomy/catalog.structs.schema.json]: Done

File: config/taxonomy/catalog.structs.schema.json (hunk 18 above)
Comment (ritwikc) [reply config/taxonomy/catalog.structs.schema.json]: Done

File: config/taxonomy/policymanager.values.schema.json (hunk 20 above)
Comment (ritwikc) [reply config/taxonomy/policymanager.values.schema.json]: Done

**Review by rohithdv — APPROVED** [review]

File: config/taxonomy/catalog.structs.schema.json (hunk 19 above)
Comment (roee88) [reply config/taxonomy/catalog.structs.schema.json:54]: Yes, I think it will be fine for this first iteration.

File: config/taxonomy/catalog.structs.schema.json [hunk 21]
```diff
@@ -28,10 +28,14 @@
             "description": "A grouping of discrete geographies that are not necessarily in a common hierarchy.",
             "items": {
                 "properties": {
-                    "geography": {"$ref": "#/definitions/geography"}
+                    "geography": { "$ref": "#/definitions/geography" }
```
[the remark this answers is not in the document]
Comment (ronenkat) [reply config/taxonomy/catalog.structs.schema.json:31]: will be fixed by #636.

**Review by ronenkat — APPROVED** [review]

**Review by simanadler — APPROVED** [review]
